### PR TITLE
feat(kuma-cp): ditch `generation` approach in finalizer and rely on CP catalog 

### PR DESCRIPTION
**Pull request by lobkovilya**
The only reason to have a `generation` counter was to detect if the CP instance is gone and the counter is not updating anymore. Now that we have Inter CP API with CP catalog we can detect this in a much simpler way.

I'm still on the edge if we should let both approaches co-exist for a few releases. But as far as I can tell there are no advantages of the `generation` approach so seems unnecessary to keep it.

### TODO

- [ ] Should we keep the `generation` approach and make a new one as `experimental`?

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues -- Fix https://github.com/kumahq/kuma/issues/6570
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/gc/finalizer.go [hunk 1]
```diff
@@ -150,32 +152,48 @@ func (f *subscriptionFinalizer) checkGeneration(ctx context.Context, typ core_mo
...
-		log.V(1).Info("mark subscription as disconnected")
-		insight.GetLastSubscription().SetDisconnectTime(now)
-
-		upsertInsight, _ := registry.Global().NewObject(typ)
-		err := manager.Upsert(ctx, f.rm, key, upsertInsight, func(r core_model.Resource) error {
-			return upsertInsight.GetSpec().(generic.Insight).UpdateSubscription(insight.GetLastSubscription())
-		})
-		if err != nil {
-			log.Error(err, "unable to finalize subscription")
+		if exist, err := f.instanceExists(ctx, insight.GetLastSubscription().GetInstanceId()); err != nil {
```
Comment (lahabana) [inline pkg/gc/finalizer.go:164]: Do we have guarantee that the `InstanceId` is always unique?

File: pkg/gc/finalizer.go [hunk 2]
```diff
@@ -150,32 +152,48 @@ func (f *subscriptionFinalizer) checkGeneration(ctx context.Context, typ core_mo
...
+			}
+			delete(f.insights[tenantID(tenantId)][typ], key)
 		}
-		delete(f.insights[tenantID(tenantId)][typ], key)
 	}
 	return nil
 }
 
+func (f *subscriptionFinalizer) instanceExists(ctx context.Context, id string) (bool, error) {
+	instances, err := f.catalog.Instances(ctx)
```
Comment (lahabana) [inline pkg/gc/finalizer.go:185]: Feels like this should be a helper in the catalog (or in the catalog package at least)

**Review by lahabana — APPROVED** [review]

File: pkg/gc/finalizer.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/gc/finalizer.go:185]:
```
func InstanceOfID(ctx context.Context, catalog Catalog, id string) (Instance, error) {
	instances, err := catalog.Instances(ctx)
	if err != nil {
		return Instance{}, err
	}
	for _, instance := range instances {
		if instance.Id == id {
			return instance, nil
		}
	}
	return Instance{}, ErrInstanceNotFound
}
```
there is this ^

**Comment by jakubdyszkiewicz** [comment]
Looks good at first glance, but I have one problem with this approach. This issue https://github.com/kumahq/kuma/issues/7467

I'm not sure we should merge it until we solve it

**Comment by lobkovilya** [comment]
@jakubdyszkiewicz I don't think #7467 will be solved in the near future, I see 2 options for how we can merge this PR:
* put catalog-based finalized under a feature flag
* make catalog-based finalizer default, but keep a way to enable generation-based finalizer if needed (i.e. when you have multiple CP replicas in multiple clusters)

**Comment by lobkovilya** [comment]
Closing while it's not clear what to do with https://github.com/kumahq/kuma/issues/7467

**Pull request closed without merging**